jiant/tasks/qa.py

```diff
@@ -441,3 +446,165 @@ def get_metrics(self, reset=False):
+            else:
+                d["sentence"] = sentence_to_text_field(
+                    model_preprocessing_interface.boundary_token_fn(sentence_tokens), indexers
+                )
+                d["question"] = sentence_to_text_field(
+                    model_preprocessing_interface.boundary_token_fn(question_tokens), indexers
+                )
+
+            d["span_start"] = NumericField(answer_span[0], label_namespace="span_start_labels")
+            d["span_end"] = NumericField(answer_span[1], label_namespace="span_end_labels")
+            d["idx"] = LabelField(idx, label_namespace="idxs", skip_indexing=True)
+            return Instance(d)
+
+        split = list(split)
+        instances = map(_make_instance, *split)
+        return instances
+
+    def _load_file(self, path):
+        example_list = []
+        aligner_fn = get_aligner_fn(self.tokenizer_name)
+        with gzip.open(path) as f:
+            lines = f.read().splitlines()
+            for line in lines:
+                datum = self.preprocess_qasrl_datum(json.loads(line))
+                sentence_tokens = datum["sentence_tokens"]
+                # " ".join because retokenizer functions assume space-delimited input tokens
+                aligner, processed_sentence_tokens = aligner_fn(" ".join(sentence_tokens))
+                for entry in datum["entries"]:
+                    for question, answer_list in entry["questions"].items():
+                        for answer in answer_list:
+                            for answer_span in answer:
+                                projected_answer_span = aligner.project_span(*answer_span["span"])
+                                # Adjust for [CLS] / <SOS> token
+                                adjusted_answer_span = (
+                                    projected_answer_span[0] + 1,
+                                    projected_answer_span[1] + 1,
+                                )
+                                example_list.append(
+                                    {
+                                        "sentence_tokens": self._process_sentence(
+                                            processed_sentence_tokens
+                                        ),
+                                        "question_tokens": self._process_sentence(question),
+                                        "answer_span": adjusted_answer_span,
+                                        "idx": len(example_list),
+                                    }
+                                )
+        return [
+            [example[k] for example in example_list]
+            for k in ["sentence_tokens", "question_tokens", "answer_span", "idx"]
+        ]
+
+    @classmethod
+    def _shuffle_data(cls, data, seed=1234):
+        # Shuffle validation data to ensure diversity in periodic validation with val_data_limit
+        indices = list(range(len(data[0])))
+        random.Random(seed).shuffle(indices)
```

Comment:
what's with the seed here? Is it not enough to seed `random` in main ?

Reply:
Is this being called somewhere right now ?

Reply:
See: https://github.com/nyu-mll/jiant/pull/716#issuecomment-506966644

The goal is to have a fixed-order shuffle of the validation data, potentially across runs. Hence, we want it to be independent of the usual random seed.